### PR TITLE
keycloak_user_federation: add mapper removal

### PR DESCRIPTION
**Pull request by fgruenbauer**
##### SUMMARY

Remove existing mappers that are not present in desired configuration for a user federation. This includes removing unwanted default mappers created on user federation creation.

Should partially fix [7169](https://github.com/ansible-collections/community.general/issues/7169)

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

keycloak_user_federation

### CODE REVIEW

File: changelogs/fragments/8695-keycloak_user_federation-mapper-removal.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - remove existing user federation mappers if they are not present in the federation configuration and will not be updated
```
Comment (felixfontein) [inline changelogs/fragments/8695-keycloak_user_federation-mapper-removal.yml:2]:
```suggestion
  - keycloak_user_federation - remove existing user federation mappers if they are not present in the federation configuration and will not be updated (https://github.com/ansible-collections/community.general/issues/7169, https://github.com/ansible-collections/community.general/pull/8695).
```

File: plugins/modules/keycloak_user_federation.py [hunk 2]
```diff
@@ -892,11 +892,11 @@ def main():
             if cid is None:
                 old_mapper = {}
             elif change.get('id') is not None:
-                old_mapper = kc.get_component(change['id'], realm)
+                old_mapper = next((before_mapper for before_mapper in before_mapper.get('mappers', []) if before_mapper["id"] == change['id']), None)
                 if old_mapper is None:
                     old_mapper = {}
             else:
-                found = kc.get_components(urlencode(dict(parent=cid, name=change['name'])), realm)
+                found = list(filter(lambda before_mapper: before_mapper['name'] == change['name'], before_comp.get('mappers', [])))
```
Comment (felixfontein) [inline plugins/modules/keycloak_user_federation.py:899]: Better use list comprehensions:
```suggestion
                found = [before_mapper for before_mapper in before_comp.get('mappers', []) if before_mapper['name'] == change['name']]
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments.

File: plugins/modules/keycloak_user_federation.py [hunk 3]
```diff
@@ -931,42 +931,51 @@ def main():
...
         cid = after_comp['id']
+        updated_mappers = []
+        # when creating a user federation, keycloak automatically creates default mappers
+        default_mappers = kc.get_components(urlencode(dict(parent=cid)), realm)
 
-        for mapper in updated_mappers:
-            found = kc.get_components(urlencode(dict(parent=cid, name=mapper['name'])), realm)
+        # create new mappers or update existing default mappers
+        for desired_mapper in desired_mappers:
+            found = list(filter(lambda default_mapper: default_mapper['name'] == desired_mapper['name'], default_mappers))
```
Comment (felixfontein) [inline plugins/modules/keycloak_user_federation.py:949]:
```suggestion
            found = [default_mapper for default_mapper in default_mappers if default_mapper['name'] == desired_mapper['name']]
```

**Comment by felixfontein** [comment]
If nobody objects, I'll merge this on Monday. Same comment as for #8735 :)

**Comment by felixfontein** [comment]
@fgruenbauer thanks for your contribution!

**Comment by fgruenbauer** [comment]
I think the only way is to add the default mappers to your argument spec. So your module argument spec would include all desired mappers, the non-standard as well as the standard.

**Comment by canon-cmre-benoit-lecardonnel** [comment]
Thanks for the quick answer.

It looks like just listing them (without config) is enough:
```yaml
    - name: Create Active Directory user federation in Keycloak
      community.general.keycloak_user_federation:
        auth_keycloak_url: "...."
        #...
        state: present
        config:
          bindDn: "..."
          #...
        mappers:
          - name: creation date
          - name: email
          - name: full name
          - name: last name
          - name: modify date
          - name: MSAD account controls
          - name: username
          - name: "my-mapper"
            providerId: "group-ldap-mapper"
            providerType: "org.keycloak.storage.ldap.mappers.LDAPStorageMapper"
            config:
              #...
```

Note that this is not ideal: I don't know if Keycloak changes the default list of mapper from version to version. If they do, this could break something silently in a future version of Keycloak if they add a mapper that's important.

Do you think `community.general.keycloak_user_federation` could get a new parameter that keeps unlisted mappers present?

Thanks,
Ben

**Comment by fgruenbauer** [comment]
You're right, the module looks for existing mappers by name or id, so the name should be enough.

Yes, that should be possible and is probably a good idea to prevent what you described. I will make a PR draft.

**Comment by canon-cmre-benoit-lecardonnel** [comment]
Just realised that the workaround I've posted above is not good enough, since it breaks when calling the module a second time (i.e. when the federation already exists, and there is effectively nothing to do).
In other words, I think the code in the `else` branch that starts at line 981 seems unable to cope with elements that only have a name.

It produces this backtrace:
```
      File "/tmp/ansible_community.general.keycloak_user_federation_payload_68l71xjj/ansible_community.general.keycloak_user_federation_payload.zip/ansible_collections/community/general/plugins/modules/keycloak_user_federation.py", line 1052, in <module>
      File "/tmp/ansible_community.general.keycloak_user_federation_payload_68l71xjj/ansible_community.general.keycloak_user_federation_payload.zip/ansible_collections/community/general/plugins/modules/keycloak_user_federation.py", line 1007, in main
      File "/tmp/ansible_community.general.keycloak_user_federation_payload_68l71xjj/ansible_community.general.keycloak_user_federation_payload.zip/ansible_collections/community/general/plugins/modules/keycloak_user_federation.py", line 1007, in <listcomp>
    KeyError: 'id'
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
```

**Comment by fgruenbauer** [comment]
Yes, i think i can reproduce the error. The problem seems to be that mappers that will be newly created dont have an id yet. 

I created PR draft [8762](https://github.com/ansible-collections/community.general/pull/8762). Could you test it @canon-cmre-benoit-lecardonnel ? 

It looks like some of your mappers are newly created (again), if thats the case it might be necessary to the add the full config to the mappers.

**Comment by canon-cmre-benoit-lecardonnel** [comment]
Thanks, I'll have a look tomorrow and let you know how it's going.

**Comment by uejo** [comment]
Hello, 
it appears the code above has an issue when defining an id in the mappers

eg:
```
mappers:
  - name: "first name"
    id: f2d82dec-923c-40da-a9f7-0fe83ca8c009
    providerId: "user-attribute-ldap-mapper"
    providerType: "org.keycloak.storage.ldap.mappers.LDAPStorageMapper"
    config:
      ldap.attribute: givenName
      read.only: true
      write.only: false
      attribute.force.default: true
      is.mandatory.in.ldap: true
      is.binary.attribute: false
      user.model.attribute: firstName
```
          
i get the stacktrace:
```
Traceback (most recent call last):
  File "/home/orchid/.ansible/tmp/ansible-tmp-1725533586.014906-48515-214085464069329/AnsiballZ_keycloak_user_federation.py", line 107, in <module>
    _ansiballz_main()
  File "/home/orchid/.ansible/tmp/ansible-tmp-1725533586.014906-48515-214085464069329/AnsiballZ_keycloak_user_federation.py", line 99, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/orchid/.ansible/tmp/ansible-tmp-1725533586.014906-48515-214085464069329/AnsiballZ_keycloak_user_federation.py", line 47, in invoke_module
    runpy.run_module(mod_name='ansible_collections.community.general.plugins.modules.keycloak_user_federation', init_globals=dict(_module_fqn='ansible_collections.community.general.plugins.modules.keycloak_user_federation', _modlib_path=modlib_path),
  File "<frozen runpy>", line 226, in run_module
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/tmp/ansible_community.general.keycloak_user_federation_payload_1m0jli33/ansible_community.general.keycloak_user_federation_payload.zip/ansible_collections/community/general/plugins/modules/keycloak_user_federation.py", line 1052, in <module>
  File "/tmp/ansible_community.general.keycloak_user_federation_payload_1m0jli33/ansible_community.general.keycloak_user_federation_payload.zip/ansible_collections/community/general/plugins/modules/keycloak_user_federation.py", line 895, in main
UnboundLocalError: cannot access local variable 'before_mapper' where it is not associated with a value
```

**Comment by fgruenbauer** [comment]
Hi @uejo, 
thanks for the report. Could you test if [8831](https://github.com/ansible-collections/community.general/pull/8831) fixes the issue for you?

**Comment by uejo** [comment]
lgtm, thanks for the fix!

**Pull request merged**